stdlib/regex.gr

```diff
@@ -0,0 +1,2946 @@
+/* Grain Regular Expression Library
+ *
+ * This library provides support for regular expressions in Grain.
+ * Its parser and analyzer are largely ported from Racket (https://racket-lang.org/),
+ * which is licensed under Apache 2.0. Racket's regular expression
+ * engine is itself inspired by the Spencer engine, as found in Tcl.
+ */
+import Array from "array"
+import Char from "char"
+import List from "list"
+import Map from "map"
+import Option from "option"
+import Result from "result"
+import String from "string"
+import Float32 from "float32"
+
+/*
+
+===============================
+REGEX PARSER CONFIG DEFINITIONS
+===============================
+
+*/
+
+record RegExParserConfig {
+  isPerlRegExp: Bool,
+  caseSensitive: Bool,
+  multiLine: Bool,
+  groupNumber: Box<Number>,
+  references: Box<Bool>,
+};
+
+let makeRegExParserConfig = () => {
+  {
+    isPerlRegExp: true,
+    caseSensitive: true,
+    multiLine: false,
+    groupNumber: box(0),
+    references: box(false),
+  }
+};
+
+let configWithCaseSensitive = (config: RegExParserConfig, caseSensitive: Bool) => {
+  {
+    isPerlRegExp: config.isPerlRegExp,
+    caseSensitive: caseSensitive,
+    multiLine: config.multiLine,
+    groupNumber: config.groupNumber,
+    references: config.references,
+  }
+};
+
+let configWithMultiLine = (config: RegExParserConfig, multiLine: Bool) => {
+  {
+    isPerlRegExp: config.isPerlRegExp,
+    caseSensitive: config.caseSensitive,
+    multiLine: multiLine,
+    groupNumber: config.groupNumber,
+    references: config.references,
+  }
+};
+
+let configGroupNumber = (config: RegExParserConfig) => unbox(config.groupNumber);
+
+let configIncGroupNumber = (config: RegExParserConfig) => {
+  config.groupNumber := unbox(config.groupNumber) + 1;
+  config
+};
+
+record RegExBuf {
+  input: String,
+  inputExploded: Array<Char>,
+  cursor: Box<Number>,
+  config: RegExParserConfig,
+};
+
+let makeRegExBuf = (s) => {
+  {input: s, inputExploded: String.explode(s), cursor: box(0), config: makeRegExParserConfig()}
+};
+
+let withConfig = (buf: RegExBuf, config: RegExParserConfig) => {
+  {input: buf.input, inputExploded: buf.inputExploded, cursor: buf.cursor, config: config}
+};
+
+// Parsing internals for recursive descent
+
+let parseErr = (buf: RegExBuf, msg: String, posShift) => {
+  "Invalid Regular Expression: " ++ msg ++ " (position " ++ toString(unbox(buf.cursor) + posShift) ++ ")"
```

Comment:
Should we have actual exceptions that people can pattern match on? Maybe something like this:
```suggestion
  RegexParseError("Invalid Regular Expression: " ++ msg ++ " (position " ++ toString(unbox(buf.cursor) + posShift) ++ ")")
```

Reply:
We can, but I'm not sure I see the benefit unless there are multiple variants, and I cannot think of multiple useful variants to define.

Reply:
For the record, I made this change locally and did not see any benefit from it. Remember that `Result`'s `Err` value is typed as well, in case you were thinking something along the lines of the open type patterns (e.g. warnings) that we use in the compiler's Reason code.

Reply:
I hear you, but do we expect that we will be adding other types of errors to this module? I honestly expect this to not really evolve after it is merged (except for the byte-regexp matching, which would not introduce any new error types).

Reply:
Resolving per offline discussion